# Error Measurements

There are several different methods to calculate the error of a measurement om comparison to a
theoretical (or approximated value):

## General Error

- "signed error" (also "error"): $$measurement - theoretical$$
- "absolute error": $$abs(measurement - theoretical)$$

## Relative Error

based on these there are several other quantities (they are possible for signed and absolute errors)

- "relative error": $$error \div theoretical$$
- "percentage error": $$100 \times error \div theoretical$$

## Squared Error

and several more advanced error definitions (based on square):

- "square_error", "square_relative_error"
- "mean_square_error" (also "mse"), "mean_square_relative_error"
- "root_mean_square_error" (also "rmse"), "root_mean_square_relative_error" (also "rmse_rel")
- "median_square_error", "median_square_relative_error"
- "sum_square_error" (also avaible under "residual_sum_of_squares" or "rss"), "sum_square_relative_error"

## Mean Error

- "mean_signed_error", "mean_absolute_error"
- "mean_relative_signed_error", "mean_relative_absolute_error"
- "mean_percentage_signed_error", mean_percentage_absolute_error"

## Median Error

- "median_signed_error", "median_absolute_error", "median_absolute_standard_deviation"
- "median_relative_signed_error", "median_relative_absolute_error"
- "median_percentage_signed_error", "median_percentage_absolute_error"

In [1]:
from pymsastro.stats import *

For simplicity I'll create a function that will call all the error functions and displays the result at the end

In [2]:
# For simplicity I'll make a function that calls each error calculation function and prints the result
def make_error_computation(measured, theory):
    error = [signed_error, absolute_error,
             relative_signed_error, relative_absolute_error,
             percentage_signed_error, percentage_absolute_error,
             square_error, square_relative_error,
             mean_square_error, mean_square_relative_error,
             root_mean_square_error, root_mean_square_relative_error,
             median_square_error, median_square_relative_error,
             sum_square_error, sum_square_relative_error,
             mean_signed_error, mean_absolute_error,
             mean_relative_signed_error, mean_relative_absolute_error,
             mean_percentage_signed_error, mean_percentage_absolute_error,
             median_signed_error, median_absolute_error, median_absolute_standard_deviation,
             median_relative_signed_error, median_relative_absolute_error,
             median_percentage_signed_error, median_percentage_absolute_error]

    for i in error:
        print("{0}: {1}".format(i.__name__, i(measured, theory)))

Now let's have some try. Suppose we have a measurement of something that has a theoretical value of 100. The value is measured once with 101.

In [3]:
theory = 100
measurement = 101

make_error_computation(measurement, theory)

signed_error: 1
absolute_error: 1
relative_signed_error: 0.01
relative_absolute_error: 0.01
percentage_signed_error: 1.0
percentage_absolute_error: 1.0
square_error: 1
square_relative_error: 0.0001
mean_square_error: 1.0
mean_square_relative_error: 0.0001
root_mean_square_error: 1.0
root_mean_square_relative_error: 0.01
median_square_error: 1.0
median_square_relative_error: 0.0001
sum_square_error: 1
sum_square_relative_error: 0.0001
mean_signed_error: 1.0
mean_absolute_error: 1.0
mean_relative_signed_error: 0.01
mean_relative_absolute_error: 0.01
mean_percentage_signed_error: 1.0
mean_percentage_absolute_error: 1.0
median_signed_error: 1.0
median_absolute_error: 1.0
median_absolute_standard_deviation: 1.482602218505602
median_relative_signed_error: 0.01
median_relative_absolute_error: 0.01
median_percentage_signed_error: 1.0
median_percentage_absolute_error: 1.0


Okay most of them are the same and very little happens here. So we measure the value again and this time it is 98.

In [4]:
theory = 100
measurement = 98

make_error_computation(measurement, theory)

signed_error: -2
absolute_error: 2
relative_signed_error: -0.02
relative_absolute_error: 0.02
percentage_signed_error: -2.0
percentage_absolute_error: 2.0
square_error: 4
square_relative_error: 0.0004
mean_square_error: 4.0
mean_square_relative_error: 0.0004
root_mean_square_error: 2.0
root_mean_square_relative_error: 0.02
median_square_error: 4.0
median_square_relative_error: 0.0004
sum_square_error: 4
sum_square_relative_error: 0.0004
mean_signed_error: -2.0
mean_absolute_error: 2.0
mean_relative_signed_error: -0.02
mean_relative_absolute_error: 0.02
mean_percentage_signed_error: -2.0
mean_percentage_absolute_error: 2.0
median_signed_error: -2.0
median_absolute_error: 2.0
median_absolute_standard_deviation: 2.965204437011204
median_relative_signed_error: -0.02
median_relative_absolute_error: 0.02
median_percentage_signed_error: -2.0
median_percentage_absolute_error: 2.0


At least this time the difference between signed and absolute errors is visible but it is still mostly the same value. But what happens if we include both measured values?

I'll use a numpy array to include both measurements.

In [5]:
import numpy as np
theory = 100
measurement = np.array([101, 98])

make_error_computation(measurement, theory)

signed_error: [ 1 -2]
absolute_error: [1 2]
relative_signed_error: [ 0.01 -0.02]
relative_absolute_error: [ 0.01  0.02]
percentage_signed_error: [ 1. -2.]
percentage_absolute_error: [ 1.  2.]
square_error: [1 4]
square_relative_error: [ 0.0001  0.0004]
mean_square_error: 2.5
mean_square_relative_error: 0.00025
root_mean_square_error: 1.5811388300841898
root_mean_square_relative_error: 0.015811388300841896
median_square_error: 2.5
median_square_relative_error: 0.00025
sum_square_error: 5
sum_square_relative_error: 0.0005
mean_signed_error: -0.5
mean_absolute_error: 1.5
mean_relative_signed_error: -0.005
mean_relative_absolute_error: 0.015
mean_percentage_signed_error: -0.5
mean_percentage_absolute_error: 1.5
median_signed_error: -0.5
median_absolute_error: 1.5
median_absolute_standard_deviation: 2.223903327758403
median_relative_signed_error: -0.005
median_relative_absolute_error: 0.015
median_percentage_signed_error: -0.5
median_percentage_absolute_error: 1.5


Okay that's more like a statistic. But since we have the experiment up and running we will start measuring the value much more often. Suppose a million times.

Instead of typing all different measurements I'll use the pseudo-random number generator of numpy with a normal distributed value of 100 and a standard deviation of 3 and we'll measure it a million times.

In [6]:
theory = 100
measurement = np.random.normal(100, 3, 1000000)

make_error_computation(measurement, theory)

signed_error: [ 1.33826562  2.21324145  1.37396214 ...,  3.2593559   0.0873033  -3.6159892 ]
absolute_error: [ 1.33826562  2.21324145  1.37396214 ...,  3.2593559   0.0873033   3.6159892 ]
relative_signed_error: [ 0.01338266  0.02213241  0.01373962 ...,  0.03259356  0.00087303
 -0.03615989]
relative_absolute_error: [ 0.01338266  0.02213241  0.01373962 ...,  0.03259356  0.00087303
  0.03615989]
percentage_signed_error: [ 1.33826562  2.21324145  1.37396214 ...,  3.2593559   0.0873033  -3.6159892 ]
percentage_absolute_error: [ 1.33826562  2.21324145  1.37396214 ...,  3.2593559   0.0873033   3.6159892 ]
square_error: [  1.79095488e+00   4.89843771e+00   1.88777197e+00 ...,   1.06234009e+01
   7.62186642e-03   1.30753779e+01]
square_relative_error: [  1.79095488e-04   4.89843771e-04   1.88777197e-04 ...,   1.06234009e-03
   7.62186642e-07   1.30753779e-03]
mean_square_error: 8.998769805333971
mean_square_relative_error: 0.0008998769805333975
root_mean_square_error: 2.9997949605487992
root_me

Now this is more like a statistic and some quantities are very handy to have.

- The ``root_mean_square_error`` is almost 3. Which is just the standard deviation of our random number generator. So this can be used as a measurement for the standard-deviation.

- The ``mean_square_error`` is the square of the ``root_mean_square_error`` and since we identified the latter as standard-deviation approximation the square of it must be the variance. A difference between the mean_squared_error and the variance would be identified as the BIAS of the measurement.

- Apart from the ``root_mean_square_error`` the ``median_absolute_error`` is also an approximator for the standard deviation but since the median has some other characteristics than the mean we have to multiply it by approximatly 1.4826 (see https://en.wikipedia.org/wiki/Median_absolute_deviation) for an explanation. This ``median_absolute_standard_deviation`` estimator is also included.

- The ``mean_absolute_difference`` is a measurement for the statistical dispersion.

- The ``mean_signed_error`` and ``median_signed_error`` are a measurement for the goodness of the measured value to the arithmetic mean or median of the sample. A value of zero or close to zero means that the mean/median of the sample is very similar to the theoretical value.

- The ``mean_percentage_error`` gives the average percentage of difference for each measurement.

- Another important quantity is the ``root_mean_square_relative_error`` since the inverse of it would give the *signal-to-noise ratio*

But another example where most statistic packages I know give strange results: If the theoretical value is negative. It will not happen often (rather very rare) but these error measurements can cope with it.

Suppose now we have a theoretical value of -100 and measure -95

In [7]:
theory = -100
measurement = -95

make_error_computation(measurement, theory)

signed_error: 5
absolute_error: 5
relative_signed_error: 0.05
relative_absolute_error: 0.05
percentage_signed_error: 5.0
percentage_absolute_error: 5.0
square_error: 25
square_relative_error: 0.0025000000000000005
mean_square_error: 25.0
mean_square_relative_error: 0.0025000000000000005
root_mean_square_error: 5.0
root_mean_square_relative_error: 0.05
median_square_error: 25.0
median_square_relative_error: 0.0025000000000000005
sum_square_error: 25
sum_square_relative_error: 0.0025000000000000005
mean_signed_error: 5.0
mean_absolute_error: 5.0
mean_relative_signed_error: 0.05
mean_relative_absolute_error: 0.05
mean_percentage_signed_error: 5.0
mean_percentage_absolute_error: 5.0
median_signed_error: 5.0
median_absolute_error: 5.0
median_absolute_standard_deviation: 7.41301109252801
median_relative_signed_error: 0.05
median_relative_absolute_error: 0.05
median_percentage_signed_error: 5.0
median_percentage_absolute_error: 5.0


Works as expected. Just to be on the safe side we measure -102 for the same quantity.

In [8]:
theory = -100
measurement = -102

make_error_computation(measurement, theory)

signed_error: -2
absolute_error: 2
relative_signed_error: -0.02
relative_absolute_error: 0.02
percentage_signed_error: -2.0
percentage_absolute_error: 2.0
square_error: 4
square_relative_error: 0.0004
mean_square_error: 4.0
mean_square_relative_error: 0.0004
root_mean_square_error: 2.0
root_mean_square_relative_error: 0.02
median_square_error: 4.0
median_square_relative_error: 0.0004
sum_square_error: 4
sum_square_relative_error: 0.0004
mean_signed_error: -2.0
mean_absolute_error: 2.0
mean_relative_signed_error: -0.02
mean_relative_absolute_error: 0.02
mean_percentage_signed_error: -2.0
mean_percentage_absolute_error: 2.0
median_signed_error: -2.0
median_absolute_error: 2.0
median_absolute_standard_deviation: 2.965204437011204
median_relative_signed_error: -0.02
median_relative_absolute_error: 0.02
median_percentage_signed_error: -2.0
median_percentage_absolute_error: 2.0


It works even if you have a theoretical negative value and measure a positive value (or vise-versa)

In [9]:
theory = -2
measurement = 1

make_error_computation(measurement, theory)

signed_error: 3
absolute_error: 3
relative_signed_error: 1.5
relative_absolute_error: 1.5
percentage_signed_error: 150.0
percentage_absolute_error: 150.0
square_error: 9
square_relative_error: 2.25
mean_square_error: 9.0
mean_square_relative_error: 2.25
root_mean_square_error: 3.0
root_mean_square_relative_error: 1.5
median_square_error: 9.0
median_square_relative_error: 2.25
sum_square_error: 9
sum_square_relative_error: 2.25
mean_signed_error: 3.0
mean_absolute_error: 3.0
mean_relative_signed_error: 1.5
mean_relative_absolute_error: 1.5
mean_percentage_signed_error: 150.0
mean_percentage_absolute_error: 150.0
median_signed_error: 3.0
median_absolute_error: 3.0
median_absolute_standard_deviation: 4.447806655516806
median_relative_signed_error: 1.5
median_relative_absolute_error: 1.5
median_percentage_signed_error: 150.0
median_percentage_absolute_error: 150.0


# Other measurement functions

There are some functions included that are not strictly about errors. These include:

- "root_mean_square" (also "rms"): The RMS of an array
- "sum_square" (also "ss", "sum_of_squares"): The sum of the squares of an array
- "root_sum_square": The square root of the sum_square

# Also different "mean" methods are imported from numpy, scipy:

- "arithmetic_mean": "numpy.mean"
- "quadratic_mean": "root_mean_square"
- "harmonic_mean": "scipy.stats.hmean"
- "geometric_mean": "scipy.stats.gmean"

For simplicity I'll make a function that calls each calculation function and prints the result like I did with the errors.

In [10]:
def make_other_computation(measured):
    other = [root_mean_square, sum_square, root_sum_square,
             arithmetic_mean, quadratic_mean, harmonic_mean, geometric_mean]

    for i in other:
        try:
            print("{0}: {1}".format(i.__name__, i(measured)))
        except IndexError:
            print("{0} cannot be used with this input.".format(i.__name__))

Since part of these functions are defined by scipy and numpy they might not accept single values:

In [11]:
measurement = 100
make_other_computation(measurement)

root_mean_square: 100.0
sum_square: 10000
root_sum_square: 100.0
arithmetic_mean: 100.0
root_mean_square: 100.0
harmonic_mean cannot be used with this input.
geometric_mean cannot be used with this input.


But all of these work with arrays

In [12]:
measurement = np.array([1,2,3])
make_other_computation(measurement)

root_mean_square: 2.160246899469287
sum_square: 14
root_sum_square: 3.7416573867739413
arithmetic_mean: 2.0
root_mean_square: 2.160246899469287
harmonic_mean: 1.6363636363636365
geometric_mean: 1.8171205928321397


and again with a different array:

In [13]:
measurement = np.array([1,2,3,3,2,1])
make_other_computation(measurement)

root_mean_square: 2.160246899469287
sum_square: 28
root_sum_square: 5.291502622129181
arithmetic_mean: 2.0
root_mean_square: 2.160246899469287
harmonic_mean: 1.6363636363636365
geometric_mean: 1.8171205928321397


# Benchmarks
Python functions cause overhead so for single elements it slows down the code (up to 15x slower)
Also the functions are optimized to allow for negative theoretical values and more, so
sometimes there are unnecessary calls to abs() which could be avoided if you can exclude that
theoretical values could be negative.

But as speed does not always matter, sometimes having more descriptive names is better in
understanding the code you have written years ago.

So using these functions is limited to cases where:
- speed does not matter
- a more descriptive function is more comprehensable than just writing the operation
- arrays are used.

- you are not always sure where the absolutes have to be (like me) ... :-)  but I'm not the only one http://mathworld.wolfram.com/RelativeDeviation.html gives the relative deviation as "abs(measured-theory)/theory" but it should be "abs((measured-theory)/theory)" if we allow theoretical values below 0... not quite often but it happens.

## Signed Error Computation difference
Error function is 7-8 times slower.

In [14]:
%timeit 100-98
%timeit signed_error(100,98)
assert 100-98 == signed_error(100,98)

The slowest run took 18.33 times longer than the fastest. This could mean that an intermediate result is being cached 
10000000 loops, best of 3: 60.9 ns per loop
The slowest run took 7.96 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 421 ns per loop


## Relative Signed Error Computation difference
Error function is 10 times slower.

In [15]:
%timeit (98-100)/100
%timeit relative_signed_error(98, 100)
import numpy as np
assert (98-100)/100 == relative_signed_error(98, 100)

10000000 loops, best of 3: 60.8 ns per loop
The slowest run took 7.45 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 749 ns per loop


## Percentage Signed Error Computation difference
Error function is 15 times slower.

In [16]:
%timeit 100*(98-100)/100
%timeit percentage_signed_error(98, 100)
assert 100*(98-100)/100 == percentage_signed_error(98, 100)

10000000 loops, best of 3: 61.3 ns per loop
The slowest run took 7.84 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 926 ns per loop


## Percentage Signed Error Computation difference with Arrays
Error function is only a little slower (even for a 1-element array) due to numpy array overhead.

In [17]:
a = np.ones(1) * 98
b = np.ones(1) * 100
%timeit 100*(a-b)/b
%timeit percentage_signed_error(a, b)

100000 loops, best of 3: 15.9 µs per loop
10000 loops, best of 3: 22.1 µs per loop
